### ISSUE TITLE
Complement usages of await operator with invokations of ConfigureAwait(false)

### ISSUE THREAD

**Issue by Deilan** [Microsoft/BotBuilder#1741]
According to MSDN Article [Async/Await - Best Practices in Asynchronous Programming](https://msdn.microsoft.com/en-us/magazine/jj991977.aspx), all usages of `await` should be complemented with usage of `ConfigureAwait(false)` - everywhere where it's applicable.

**Comment by willportnoy** [comment]
Because we call into the user's methods within an IDialog-derived class, we did not want to non-deterministically lose any SynchronizationContext the user was expecting.

Because we don't synchronously wait on any tasks, we should not deadlock.

One remaining risk is excess latency for scheduling task continuations, subject to any throttling inherent in the ambient SynchronizationContext.

Of course, the caller of the library is free to drop the synchronization context before calling into the library if they feel it's a concern.

**Comment by Deilan** [comment]
Thanks for the answer! Would you kindly also explain when ConfigureAwait(false) should **not** be used in one's bot code? When it may lead into issues? Specifically in context of using Bot Framework.

**Comment by willportnoy** [comment]
If one of your bot methods depended on the static HttpContext.Current, then using ConfigureAwait(false) beforehand might prevent you from getting the HttpContext from that static variable.

But yes, overall (e.g. in non-UI code), using ConfigureAwait is a good practice.

**Comment by Deilan** [comment]
Thanks for the info!

Well, I think this is a pretty rare case, so losing performance just because of that is a quite big price.

One more thought about my initial suggestion about using `.ConfigureAwait(false)`: maybe provide some sort of configuration option, which will control this behavior? If the option is enabled, apply .ConfigureAwait(false) . If it is disabled, behave as now. And make this option disabled by default (for safety). I know it's not simple to implement, for now it's just a general idea on how to resolve this issue.

**Comment by willportnoy** [comment]
Before calling into the builder library, you can try to drop the SynchronizationContext and that might accomplish the same goal.

I'm going to close this issue, but feel free to open a new issue for further discussion.
